drivers/lora/sx126x.c

```diff
@@ -0,0 +1,475 @@
+
+	LOG_DBG("Issuing opcode 0x%x w. %" PRIu16 " bytes of data",
+		opcode, size);
+	sx126x_spi_transceive(req, NULL, sizeof(req), buffer, NULL, size);
+}
+
+void SX126xReadBuffer(uint8_t offset, uint8_t *buffer, uint8_t size)
+{
+	uint8_t req[] = {
+		RADIO_READ_BUFFER,
+		offset,
+		0x00,
+	};
+
+	LOG_DBG("Reading buffers @ 0x%" PRIx8 " (%" PRIu8 " bytes)",
+		offset, size);
+	sx126x_spi_transceive(req, NULL, sizeof(req), NULL, buffer, size);
+}
+
+void SX126xWriteBuffer(uint8_t offset, uint8_t *buffer, uint8_t size)
+{
+	uint8_t req[] = {
+		RADIO_WRITE_BUFFER,
+		offset,
+	};
+
+	LOG_DBG("Writing buffers @ 0x%" PRIx8 " (%" PRIu8 " bytes)",
+		offset, size);
+	sx126x_spi_transceive(req, NULL, sizeof(req), buffer, NULL, size);
+}
+
+void SX126xAntSwOn(void)
+{
+#if HAVE_GPIO_ANTENNA_ENABLE
+	LOG_DBG("Enabling antenna switch");
+	gpio_pin_set(dev_data.antenna_enable, GPIO_ANTENNA_ENABLE_PIN, 1);
+#else
+	LOG_DBG("No antenna switch configured");
+#endif
+}
+
+void SX126xAntSwOff(void)
+{
+#if HAVE_GPIO_ANTENNA_ENABLE
+	LOG_DBG("Disabling antenna switch");
+	gpio_pin_set(dev_data.antenna_enable, GPIO_ANTENNA_ENABLE_PIN, 0);
+#else
+	LOG_DBG("No antenna switch configured");
+#endif
+}
+
+uint32_t SX126xGetBoardTcxoWakeupTime(void)
+{
+	return TCXO_POWER_STARTUP_DELAY_MS;
+}
+
+uint8_t SX126xGetDeviceId(void)
+{
+	return SX126X_DEVICE_ID;
+}
+
+void SX126xIoIrqInit(DioIrqHandler dioIrq)
+{
+	LOG_DBG("Configuring DIO IRQ callback");
+	dev_data.radio_dio_irq = dioIrq;
+}
+
+void SX126xIoTcxoInit(void)
+{
+#if HAVE_DIO3_TCXO
+	CalibrationParams_t cal = {
+		.Value = SX126X_CALIBRATION_ALL,
+	};
+
+	LOG_DBG("TCXO on DIO3");
+
+	/* Delay in units of 15.625 us (1/64 ms) */
+	SX126xSetDio3AsTcxoCtrl(TCXO_DIO3_VOLTAGE,
+				TCXO_POWER_STARTUP_DELAY_MS << 6);
+	SX126xCalibrate(cal);
+#else
+	LOG_DBG("No TCXO configured");
+#endif
+}
+
+void SX126xReset(void)
+{
+	LOG_DBG("Resetting radio");
+	gpio_pin_set(dev_data.reset, GPIO_RESET_PIN, 1);
+	k_sleep(K_MSEC(20));
+	gpio_pin_set(dev_data.reset, GPIO_RESET_PIN, 0);
+	k_sleep(K_MSEC(10));
+}
+
+void SX126xSetRfTxPower(int8_t power)
+{
+	LOG_DBG("power: %" PRIi8, power);
+	SX126xSetTxParams(power, RADIO_RAMP_40_US);
+}
+
+void SX126xWaitOnBusy(void)
+{
+	while (gpio_pin_get(dev_data.busy, GPIO_BUSY_PIN)) {
+		k_sleep(K_MSEC(1));
+	}
+}
+
+void SX126xWakeup(void)
+{
+	int ret;
+
+	uint8_t req[] = { RADIO_GET_STATUS, 0 };
+	const struct spi_buf tx_buf = {
+		.buf = req,
+		.len = sizeof(req),
+	};
+
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+
+	LOG_DBG("Sending GET_STATUS");
+	ret = spi_write(dev_data.spi, &dev_data.spi_cfg, &tx);
+	if (ret < 0) {
+		LOG_ERR("SPI transaction failed: %i", ret);
+		return;
+	}
+
+	LOG_DBG("Waiting for device...");
+	SX126xWaitOnBusy();
+	LOG_DBG("Device ready");
+}
+
+static void sx126x_dio1_irq_work_handler(struct k_work *work)
+{
+	LOG_DBG("Processing IRQ");
+	if (!dev_data.radio_dio_irq) {
+		LOG_WRN("Radio IRQ before setting radio_dio_irq.");
```

Comment:
Can we say, `DIO1 IRQ callback not defined` instead?

Reply:
Rephrased as "DIO1 interrupt without valid HAL IRQ callback.".